### PR TITLE
Use `weak_ptr<ucxx::Endpoint>` in `ErrorCallbackData`

### PR DESCRIPTION
**Pull request by pentschev**
A race could cause the `Endpoint::errorCallback` to dereference an already closed and destroyed `ucxx::Endpoint`, leading to segmentation faults and sometimes deadlocks. With the use of a `weak_ptr<Endpoint>` that can be prevented by returning before dereferencing the pointer which should be safe since the owner has already been destroyed anyway.

### CODE REVIEW

File: cpp/src/endpoint.cpp
Comment (wence-) [inline cpp/src/endpoint.cpp:129]: With my suggestion above, perhaps this would then be:

```
auto ep = std::shared_ptr<Endpoint>(new Endpoint(worker, &params, endpointErrorHandling));
ep.setCallback(...);
```

But, I realise this has a problem, because we pass `params` into `ucp_ep_create` and if we only set the error callback later, then it won't be hooked up.

Should this then be a three-phase operation:

```
auto ep = std::make_shared<Endpoint>(new Endpoint(worker, &params, endpointErrorHandling));
ep.setCallback(...);
ep.create();
```
WDYT?

It's a bigger refactor than we have here though, and still leaves the option of forgetting one of the steps. So perhaps it's not worth it now.

File: cpp/src/endpoint.cpp
Comment (wence-) [inline cpp/src/endpoint.cpp:595]: *issue*: [`std::weak_ptr::lock`](https://en.cppreference.com/w/cpp/memory/weak_ptr/lock) is `noexcept`, so this error handling isn't doing what you expect.

You probably want:
```c++
if (!(endpoint = data.endpoint.lock())) {
  return;
}
```

i.e. attempt to upgrade to a `shared_ptr` and return if that is unsuccesful.

**Review by wence- — CHANGES_REQUESTED** [review]
Some non-blocking suggestions that may be not worth it, and one fix, since `std::weak_ptr::lock` is `noexcept`.

File: cpp/src/endpoint.cpp
Comment (pentschev) [reply cpp/src/endpoint.cpp]: Thanks for catching that. This solution works too, but I think the "proper" C++ solution is `try...catch` with the `weak_ptr` being passed to the `shared_ptr` constructor, so I've done that in https://github.com/rapidsai/ucxx/pull/227/commits/454762c0a03a5b2642f24fdeaa6115d6ec85696a.

File: cpp/src/endpoint.cpp
Comment (pentschev) [reply cpp/src/endpoint.cpp]: This is certainly not what I would like to have but I agree with you it's a better implementation. We can even avoid the need for `setCallback()`, a `create()` method is sufficient. See https://github.com/rapidsai/ucxx/pull/227/commits/b36ed91c18bc883d5c9ad13b09a374aab2a595bb .

File: cpp/src/endpoint.cpp
Comment (pentschev) [reply cpp/src/endpoint.cpp]: I think this should also be addressed by https://github.com/rapidsai/ucxx/pull/227/commits/b36ed91c18bc883d5c9ad13b09a374aab2a595bb .

File: cpp/src/endpoint.cpp
Comment (pentschev) [reply cpp/src/endpoint.cpp:81]: I agree the name has become misleading, I initially added it exclusively as means to communicate with the error callback, but now it is used to store any information that _may_ require to be shared with the error callback, such as the `_callbackData.closing` flag which is used in various parts of the code inside and out of the `Endpoint::errorCallback()`. I think we could handle it in two ways:

1. Rename `ErrorCallbackData` to something that is more appropriate (do you have any naming suggestions?); or
1. Duplicate things like `_callbackData.closing`, but I personally hate this because it's certain to become a problem when someone makes a change that accidentally missing synchronizing.

File: cpp/src/endpoint.cpp
Comment (pentschev) [reply cpp/src/endpoint.cpp:78]: This should also be addressed by https://github.com/rapidsai/ucxx/pull/227/commits/b36ed91c18bc883d5c9ad13b09a374aab2a595bb .

File: cpp/src/endpoint.cpp
Comment (wence-) [reply cpp/src/endpoint.cpp:81]: tl;dr: maybe a name change is OK if we can think of a good one, otherwise, do nothing,

Of the two, I definitely (like you) don't like (2). We could just call it `EndpointData`? But then the question arises, in what circumstances do I need this data where I don't have the endpoint? IOW, if the `Endpoint` directly provided the information that we need to pass in to `ucp_ep_create` (say), could we just use it directly? Or does that bring all sorts of problems with destruction order etc...

**Review by wence- — APPROVED** [review]
Thanks Peter! I'll let you decide whether or not you think it's worthwhile renaming `errorcallbackdata`

File: cpp/src/endpoint.cpp
Comment (pentschev) [reply cpp/src/endpoint.cpp:81]: > But then the question arises, in what circumstances do I need this data where I don't have the endpoint?

This really shouldn't be necessary as the `ErrorCallbackData` is itself owned by the `Endpoint`. I haven't been able to determine why it may happen at times that the endpoint looks like it's been destroyed, but I'm thinking it may have something to do with generic callbacks.

I think one of the issues where we're seeing segfaults elsewhere are the generic callbacks, specifically after a timeout occurred, the [`Endpoint::closeBlocking`](https://github.com/rapidsai/ucxx/blob/b117fda698c76a3685b5b7d14008609eb8d1d8a9/cpp/src/endpoint.cpp#L226-L243) is one such example, we register the callback with the worker and wait for it, if it doesn't complete we simply get to the next iteration of the loop. The problem is those objects are all locally-scoped and we never deregister from the worker, so if/when the worker actually attempts to execute the callback the function and objects do not exist anymore, leading to potential segfaults, which is definitely something I'll try to address in a follow-up PR within the next few days.

> IOW, if the `Endpoint` directly provided the information that we need to pass in to `ucp_ep_create` (say), could we just use it directly? Or does that bring all sorts of problems with destruction order etc...

This is an excellent point you're making, we don't really need the `ErrorCallbackData` provided we can pass a pointer to `this` directly to `errorCallback()`, which we now can do thanks to the `create()` method. With that said I've now attempted to remove `ErrorCallbackData` and pass `this` directly to a `errorCallback` function which is now a `friend` of `ucxx::Endpoint` and can thus operate directly on the `Endpoint` instance itself. However, I'm not confident this won't cause the original issue I'm trying to address with this PR so before we merge it I'll have a stress test running locally for a day or so and will trigger CI multiple times to see if we can reproduce the segfault.

Comment (pentschev) [reply cpp/src/endpoint.cpp:123]: The change is in https://github.com/rapidsai/ucxx/pull/227/commits/ade6aff99c906f3fafdcf11120ab66d82e6b7360 .

Comment (pentschev) [reply cpp/src/endpoint.cpp:81]: As discussed offline, unfortunately https://github.com/rapidsai/ucxx/commit/ade6aff99c906f3fafdcf11120ab66d82e6b7360 has reintroduced the segfault which is likely rooted at the generic callbacks, therefore I reverted it and will reintroduced that change along with generic callback fixes in a follow-up PR.

**Comment by pentschev** [comment]
Thanks @wence- for another insightful review!

**Pull request merged**
